course/src/Vector3.cc

```diff
@@ -0,0 +1,156 @@
+#include "Vector3.h"
+
+namespace ekumen
+{
+
+Vector3::Vector3(const double x, const double y, const double z)
+{
+    values[0] = x;
+    values[1] = y;
+    values[2] = z;
+}
+Vector3::Vector3()
+{
+    values[0] = 0.;
+    values[1] = 0.;
+    values[2] = 0.;
+}
+Vector3::Vector3(const std::initializer_list<double> &list){
+    values[0] = *(list.begin() + 0);
+    values[1] = *(list.begin() + 1);
+    values[2] = *(list.begin() + 2);
+}
+double Vector3::norm() const
+{
+    return sqrt(dot(Vector3(x(), y(), z())));
+}
+
+double Vector3::dot(const Vector3 &q) const
+{
+    return ((values[0] * q.x()) + (values[1] * q.y()) + (values[2] * q.z()));
+}
+
+Vector3 Vector3::cross(const Vector3 &q) const
+{
+    Vector3 r;
+    r.x() = values[1] * q.z() - values[2] * q.y();
+    r.y() = values[2] * q.x() - values[0] * q.z();
+    r.z() = values[0] * q.y() - values[1] * q.x();
+    return (r);
+}
+
+Vector3 &Vector3::operator+=(const Vector3 q){
+    x() += q.x();
+    y() += q.y();
+    z() += q.z();
+    return *this;
+}
+
+Vector3 &Vector3::operator-=(const Vector3 q){
+    x() -= q.x();
+    y() -= q.y();
+    z() -= q.z();
+    return *this;
+}
+
+Vector3 operator+(Vector3 p, Vector3 q)
+{   
+    Vector3 r;    
+    r.x() = p.x();
+    r.y() = p.y();
+    r.z() = p.z();
+    return (r+=q);
+}
+
+Vector3 operator-(Vector3 p, Vector3 q)
+{
+    Vector3 r;    
+    r.x() = p.x();
+    r.y() = p.y();
+    r.z() = p.z();
+    return (r-=q);
+}
+
+Vector3 operator*(const double &cte, const Vector3 &p)
+{
+    Vector3 res(cte * p.x(), cte * p.y(), cte * p.z());
+    return res;
+}
+
+Vector3 operator*(const Vector3 &p, const double &cte)
+{
+    Vector3 res(cte * p.x(), cte * p.y(), cte * p.z());
+    return res;
+}
+
+Vector3 operator*(const Vector3 &p, const Vector3 &q)
+{
+    Vector3 res(p.x() * q.x(), p.y() * q.y(), p.z() * q.z());
+    return res;
+}
+
+Vector3 operator/(const Vector3 &p, const Vector3 &q)
+{
+    Vector3 res(p.x() / q.x(), p.y() / q.y(), p.z() / q.z());
+    return res;
+}
+
+bool operator==(const Vector3 &p, const Vector3 &q)
+{
+    bool res = (p.x() == q.x()) && (p.y() == q.y()) && (p.z() == q.z());
+    return res;
+}
+
+bool operator==(const Vector3 &p, const std::initializer_list<double> &list){
+    bool res = (p.x() == *(list.begin() + 0)) && (p.y() == *(list.begin() + 1)) && (p.z() == *(list.begin() + 2));
+    return res;
+}
+
+bool operator!=(const Vector3 &p, const Vector3 &q)
+{
+
+    return (!(p == q));
+}
+
+bool operator!=(const Vector3 &p, const std::initializer_list<double> &list){
+    Vector3 q(*(list.begin() + 0),*(list.begin() + 1),*(list.begin() + 2));
```

Comment:
To do this you need to call `begin()` three times. Better call it only once and reuse the same iterator.